lib/Sema/TypeCheckStmt.cpp

```diff
@@ -377,6 +306,1039 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
+          // subspace contains an empty, then the whole space is empty.
+          SmallVector<Space, 4> simplifiedSpaces;
+          std::transform(getSpaces().begin(), getSpaces().end(),
+                         std::back_inserter(simplifiedSpaces),
+                         [](const Space &el) {
+            return el.simplify();
+          });
+          for (auto &el : simplifiedSpaces) {
+            if (el.isEmpty()) {
+              return Space();
+            }
+          }
+          return Space(Ty, Head, simplifiedSpaces);
+        }
+        case SpaceKind::Type: {
+          // If the decomposition of a space is empty, the space is empty.
+          if (canDecompose(Ty)) {
+            SmallVector<Space, 4> ss;
+            decompose(Ty, ss);
+            if (ss.empty()) {
+              return Space();
+            }
+            return *this;
+          } else {
+            return *this;
+          }
+        }
+        case SpaceKind::Disjunct: {
+          // Simplify each disjunct.
+          SmallVector<Space, 4> simplifiedSpaces;
+          std::transform(Spaces.begin(), Spaces.end(),
+                         std::back_inserter(simplifiedSpaces),
+                         [](const Space &el){
+            return el.simplify();
+          });
+          // If the disjunct is singular, unpack it into its component.
+          if (simplifiedSpaces.size() == 1) {
+            return simplifiedSpaces.front();
+          }
+
+          // Otherwise, remove any empties.
+          SmallVector<Space, 4> compatifiedSpaces;
+          std::copy_if(simplifiedSpaces.begin(), simplifiedSpaces.end(),
+                       std::back_inserter(compatifiedSpaces),
+                       [&](const Space &el) {
+            return !el.isEmpty();
+          });
+          // If the disjunct was all empty, the space is empty.
+          if (compatifiedSpaces.empty()) {
+            return Space();
+          }
+          // Else if the disjunct is singular, unpack it into its component.
+          if (compatifiedSpaces.size() == 1) {
+            return compatifiedSpaces.front();
+          }
+          return Space(compatifiedSpaces);
+        }
+        default:
+          return *this;
+        }
+      }
+
+      // Decompose a type into its component spaces.
+      static void decompose(Type tp, SmallVectorImpl<Space> &arr) {
+        assert(canDecompose(tp) && "Non-decomposable type?");
+
+        if (tp->isBool()) {
+          arr.push_back(Space(true));
+          arr.push_back(Space(false));
+        } else if (auto *E = tp->getEnumOrBoundGenericEnum()) {
+          // Look into each case of the enum and decompose it in turn.
+          auto children = E->getAllElements();
+          std::transform(children.begin(), children.end(),
+                         std::back_inserter(arr), [&](EnumElementDecl *eed) {
+            // FIXME: This shouldn't happen.
+            if (!eed->hasInterfaceType()) {
```

Comment:
If you replace this return with an `llvm_unreachable`, then `test/NameBinding/reference-dependencies.swift` hits it while type-checking `lookUpManyTopLevelNames`.